src/engine/BMGame.php

```diff
@@ -1792,14 +1796,66 @@ protected function update_game_state_commit_attack() {
      * Perform the logic required at BMGameState::choose_turbo_swing
      */
     protected function do_next_step_choose_turbo_swing() {
+        $attacker = $this->playerArray[$this->attack['attackerPlayerIdx']];
+        foreach ($attacker->activeDieArray as $activeDie) {
+            if ($activeDie->has_skill('Turbo') &&
+                $activeDie->has_flag('IsAttacker') &&
+                $activeDie->doesReroll) {
+                $attacker->waitingOnAction = TRUE;
+            }
+        }
+    }
+
+    public function set_turbo_sizes(array $turboSizeArray) {
+        if (empty($turboSizeArray)) {
+            return TRUE;
+        }
+
+        if (BMGameState::CHOOSE_TURBO_SWING != $this->gameState) {
+            $this->message = 'Wrong game state to choose turbo swing sizes.';
+            return FALSE;
+        }
 
+        $playerIdx = $this->attack['attackerPlayerIdx'];
+        $nDice = count($this->playerArray[$playerIdx]->activeDieArray);
+        foreach (array_keys($turboSizeArray) as $key) {
+            if (!is_int($key) || ($key < 0)) {
+                $this->message = 'The turbo size array must be keyed with non-negative integer die indices';
+                return FALSE;
+            }
+
+            if ($key > $nDice - 1) {
+                $this->message = 'The turbo size array seems to refer to a die index that does not exist';
+                return FALSE;
+            }
+        }
+
+        // set turbo sizes
+        foreach ($turboSizeArray as $dieIdx => $dieSize) {
+            $die = $this->playerArray[$playerIdx]->activeDieArray[$dieIdx];
+            if (isset($die->swingType) &&
+                $die->has_skill('Turbo') &&
+                $die->has_flag('IsAttacker')) {
```

Comment:
I think we're going to run into trouble here with Morphing Turbo dice, which may become non-swing dice that just happen to still have Turbo.

I've also got to think more about whether we really require the player to input Turbo values in this case, where the new swing value will just be ignored.